### PR TITLE
[DebugInfo] Update policy for when to merge locations

### PR DESCRIPTION
**Pull request by SLTozer**
Following discussions on PR #114231 this patch changes the policy on merging locations, making the rule that new instructions should use a merge of the locations of all the instructions whose output is produced by the new instructions; in the case where only one instruction's output is produced, as in most InstCombine optimizations, we use only that instruction's location.

Happy to take suggestions on the wording from this one - I've not found an up-front explanation that makes it totally clear, but the examples hopefully help to clarify.

### CODE REVIEW

**Review by dwblaikie — APPROVED** [review]
SGTM, but please wait for one or two other approvals/reviews

File: llvm/docs/HowToUpdateDebugInfo.rst
Comment (felipepiovezan) [inline llvm/docs/HowToUpdateDebugInfo.rst:85]: >  with one or more new instructions

The new text is covering "new instructions" (plural!) but AFAICT we don't have an example of that in the text, which would help clarify something confusing in the text:

> The API to use is ``Instruction::applyMergedLocation``
> and the new location should be a merge of the locations
> of all the instructions whose output is produced in the
> new instructions

We talk about **the** location, but if we are in the plural case there would be possibly different merged locations, right?

I think a more precise wording is:

The API to use is ``Instruction::applyMergedLocation``. For each new instruction `I`, its new location should be a merge of the locations of all instructions whose output is produced by `I`.

**Comment by jmorse** [comment]
Sounds like a good plan to me, wording seems fine when the existing feedback is included.

**Review by jryans — COMMENTED** [review]
Looks good to me overall, so should be ready to go once other comments have been resolved. Thanks for working on these details.

**Comment by SLTozer** [comment]
I've added an example for multiple merges - it's not something that actually happens in the compiler right now, but primarily I think because DILocations aren't handled correctly through a lot of vectorizer operations (vector adds seem to just take a location from one of the input adds, for example). There is an actual example in `AMDGPULibCalls::fold_sincos`, where we create several instructions to replace separate `sin` and `cos` instructions with a merged `sincos` call, but the logic is a little obtuse for a small example imo.

**Review by adrian-prantl — APPROVED** [review]

**Review by felipepiovezan — APPROVED** [review]

**Pull request merged**
